third_party/lua-yaml/lyaml.cc

```diff
@@ -682,8 +684,10 @@ static int dump_node(struct lua_yaml_dumper *dumper)
    case MP_STR:
       str = field.sval.data;
       len = field.sval.len;
+      dval = fpconv_strtod(str, &endptr);
+
       if (yaml_is_null(str, len) || yaml_is_bool(str, len, &unused) ||
-          lua_isnumber(dumper->L, -1)) {
+          lua_isnumber(dumper->L, -1) || (isinf(dval) && endptr == str + len)) {
```

Comment:
I propose to wrap the check into the `yaml_is_number` function and call `fpconv_strtod` only when the `lua_isnumber` check fails.

It is OK to accept any value that is accepted by `fpconv_strtod`, not only a large one that is converted to `inf`. It would make the idea simpler, IMHO, and would require less explanation in comments.

I also propose to explain why and how this function works in comments. Key points:

* The idea is to catch all the string values that may be interpreted as a number by some decoder. It is OK to have a kind of false-positives: it means that these values are quoted on encoding that is harmless.
* Our decoder interprets unquoted values that are accepted by `strtoll`, `strtoull` and `strtod` as numbers. We must quote all such values in the encoder.
* `lua_isnumber` seems enough on the first glance, but it discards values with a large exponent, which are interpreted as `inf` in our decoder.
* `fpconv_strtod` seems enough on the first glance, but it doesn't accept binary `0b<...>` literals and since they're quoted by older tarantool versions it is undesirable to leave them unquoted in the newer ones.
* The best solution would be a format check without taking into account how large the significand and the exponent. (It can check first N characters for performance and simplicity, because false-positives are OK to some extent.) It needs to implement a parser that supports binary/dec/hex/(octal?) integral/floating point number literals in plain/scientific notations, so it is postponed.

Reply:
That'll be quite consistent. Added `yaml_is_number` (that works the same way yaml is decoded). Specified the difference from the YAML standard in the docs.